feat: external-secrets upgrade plus cluster/team wide secrets

Morriz opened the pull request:
## Checklist

- [ ] Architecture Design Records have been added as `adr/*.md` and appended to list in `adr/_index.md`, if applicable.
- [x] The `values-schema.yaml` file and `test/**` fixtures have been updated to reflect code changes, if applicable.
- [x] The OpenApi Schema from redkubes/otomi-api project is compatible with definitions from `values-schema.yaml` file, if applicable.
- [x] Helm releases are meeting otomi's baseline security policies, if applicable.
- [ ] Helm chart and helmfile changes are tested against upgrade scenario, if applicable.

staticvoid255 commented:
Really like the cluster wide secrets, I also see Vault has been updated.

One question - is this compatible with secrets created with older versions of the ES operator?

j-zimnowoda commented on values-schema.yaml:
```diff
@@ -822,8 +822,19 @@ definitions:
         readOnly: true
         type: string
+      clusterWide:
+        description: Will create the secret in all namespaces.
```
I can hardly find a use case for that. Wouldn't be better to say allTeamsWide ?

Morriz commented:
> Really like the cluster wide secrets, I also see Vault has been updated.
> 
> One question - is this compatible with secrets created with older versions of the ES operator?

technically not, and those secrets will be erased and recreated during the deploy time window, so any running workloads won't be affected. 
Team workloads (re-)starting during that window could be affected, but that is very unlikely and controllable. So either we announce the window and lock activity or we do it after working hours.

j-zimnowoda commented:
Thank you for you hard work and showing us new horizons. However, this PR does not adhere to PR guideline and goes far beyond the initial issue scope. I kindly ask you to split the change request into smaller chunks, so we can perform a quality review and have possibility to choose when a given feature is merged. 

The following split is advised:

- The changes to knative are part of the https://app.zenhub.com/workspaces/dev-5e7e84dddc966f05a627a19b/issues/redkubes/unassigned-issues/457 issue
- the `old` upgrade a separate PR (if it is prerequisite to this PR then we will merge it first)
- team/cluster scoped secrets as a separate PR 
- external-secrets upgrade as a separate PR

Please make sure that you inform reviewers about each problem you are solving, to avoid unnecessary questions.
Thanks

Morriz commented:
> The following split is advised:
> 
> * The changes to knative are part of the https://app.zenhub.com/workspaces/dev-5e7e84dddc966f05a627a19b/issues/redkubes/unassigned-issues/457 issue

removed and now in new PR: #968 

> * the `olm` upgrade a separate PR (if it is prerequisite to this PR then we will merge it first)

removed and now in #969 

> * team/cluster scoped secrets as a separate PR
> * external-secrets upgrade as a separate PR

These can go together as new use cases are only additions and not interfering. Please look again.

Morriz commented:
The olm PR should go first as that it is now a hit and miss scenario with `master` tag

j-zimnowoda commented on charts/team-ns/templates/external-secrets.yaml:
```diff
@@ -3,49 +3,80 @@
 {{- range $s := $v.secrets }}
 {{- $ns := index $s "namespace" | default $.Release.Namespace }}
-{{- $pathPrefix := printf "/secret/data/teams/team-%s" $v.teamId }}
+{{- $pathPrefix := printf "secret/data/teams/team-%s" $v.teamId }}
```
Hashicorp Vault is strict about paths. Are you sure that the `/` is not needed ?

Morriz replied:
They changed it and now it is without first slash. Luckily I found out rather quickly.

j-zimnowoda commented on charts/team-ns/templates/external-secrets.yaml:
```diff
@@ -3,49 +3,80 @@
+{{- $teamNamespacesCollection := print "team-" (include "helm-toolkit.utils.joinListWithSep" (dict "list" $v.teamIds "sep" ",team-")) }}
+
+apiVersion: external-secrets.io/v1beta1
+{{- if and (eq $v.teamId "admin") (or ($s.clusterWide | default false) ($s.teamWide | default false)) }}
```
Instead of constantly checking a a teamId is admin. Would it b e possible to simplify logic

```
{{ if teamId == admin }}
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret | ClusterExternalSecret
{{ else }}
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret 
{{end}}

```

Morriz replied:
It seems the same to me and your proposal is less DRY. Can we move on?

j-zimnowoda commented on charts/team-ns/templates/external-secrets.yaml:
```diff
@@ -3,49 +3,80 @@
+{{- end }}
 metadata:
+  {{- if and (eq $v.teamId "admin") (or ($s.clusterWide | default false) ($s.teamWide | default false)) }}
+    {{- if $s.clusterWide | default false }}
```
The use case for team-wide secrets seems to be very helpful as teams can have a shared pullSecret 🚀 

I am wondering if cluster-wide secret is needed, because there is already `globalPullSecret` implemented.
The `cluster-wide` secret  expands team-admin role to platform-admin role that can inject secrets to the platform namespaces. We have never considered team-admin to have any influence on platform namespaces. What do you think about it?

Morriz replied:
I means more endpoints can be added. Some teams could use both dockerhub as well as github artifacts, and change platform images to use their own baked ones.

Morriz replied:
But I do agree that the role should be otomi-admin for this, which should be added to api soon, as it is really easy to do. Until then I am fine shipping it like this. It introduces no security vulnerability, because if a team-admin was given a secret to do this, then he already has access and should be able to do so anyway. Now all other team-admins have access to that secret as well...so maybe turn it off until then?

I also will then allow a team admin to specify a list of teams instead of just checking "team wide".

j-zimnowoda replied:
> I means more endpoints can be added. Some teams could use both dockerhub as well as github artifacts, and change platform images to use their own baked ones. 
I wouldn't be worried about that use case until we hear that someone really, really needs it - YAGNI and KISS

j-zimnowoda replied:
Resolving it for now. We will observe if the cluster-wide secrets are  used and then make decision.

staticvoid255 commented:
I will test this and #977 together, here are some initial test cases, please feel free to suggest your own:

1. Successfully connects to Vault
2. Team secrets are successfully synced to relevant namespace
3. Cluster-wide secrets are visible by all teams
4. Sync is not flaky
5. Existing secrets in Vault are unaffected
6. Do existing Otomi secrets (i.e. refs to Vault secrets) require re-creation?
7. _Upon deleting an Otomi secret created with the old version, will it delete the **old** `ExternalSecret` CR for that secret, and the corresponding k8s secret?_ (maybe we can delete all old ExternalSecret CR's during upgrade process)
8. **If 7 is untrue**, will this version of the operator overwrite secrets created with the old operator? (i.e. in the event that the old secrets are not cleaned up, and you are trying to create a new secret with the same name as and an old secret in the same namespace)

staticvoid255 commented:
As far as I know, the new version of this operator allows for the declaration of path-based routing of secrets to namespaces, meaning we would no longer have to manually register each secret individually in the console, instead registering just the path prefix with a particular namespace. This would allow for easy support for all KMS's supported by the operator's SecretStore CRD, and would also provide a generic approach to storing Vault unseal keys. So in the future we can discuss adopting this kind of flow. 

For the purpose of this PR though, I'll limit testing scope solely to fixing the flaky syncing, and testing upgrade and cleanup behaviour.

staticvoid255 commented:
Also, this may be of some use as it's a migration tool for kes-to-eso, which we could run during upgrade steps - https://github.com/external-secrets/kes-to-eso

j-zimnowoda commented:
> As far as I know, the new version of this operator allows for the declaration of path-based routing of secrets to namespaces, meaning we would no longer have to manually register each secret individually in the console, instead registering just the path prefix with a particular namespace. This would allow for easy support for all KMS's supported by the operator's SecretStore CRD, and would also provide a generic approach to fetching Vault unseal keys. So in the future we can discuss adopting this kind of flow.
> 
> For the purpose of this PR though, I'll limit testing scope solely to fixing the flaky syncing, and testing upgrade and cleanup behaviour.

wow, that is interesting !

staticvoid255 reviewed (changes requested):
Hi there,

Helm release for the new operator gets stuck in pending-install state while upgrading from 0.16.21. Probably due to CRD conflicts between old and new operator and/or ServiceAccount permissions. Please fix and I'll continue testing.
<img width="1626" alt="Screenshot 2022-12-15 at 10 29 55" src="https://user-images.githubusercontent.com/59235486/207823764-0c14144a-bdd9-45b4-af11-370e883ac4fd.png">

j-zimnowoda commented on values-schema.yaml:
```diff
@@ -822,8 +821,19 @@ definitions:
+        description: Will create the secret in all namespaces.
+        type: boolean
+        default: false
+      teamWide:
```
It would be better to call it `allTeams` as teamWide suggests it is for one team.

j-zimnowoda commented on values-schema.yaml:
```diff
@@ -822,8 +821,19 @@ definitions:
       id:
         readOnly: true
         type: string
+      clusterWide:
```
It seems like clusterWide and teamWidea are mutually exclusive.

staticvoid255 commented:
Hi Maurice,

Is it intended behaviour that secrets created in team-admin be cluster-wide? 
I would recommend creating a flag or otherwise requiring the user specify that a given secret is intended as cluster-wide.
I also noticed that ExternalSecrets already created in team-admin _before_ upgrading are **not** recreated as ClusterExternalSecrets.

Sync of old secrets is broken once the operator is upgraded:
<img width="1446" alt="Screenshot 2022-12-19 at 09 51 04" src="https://user-images.githubusercontent.com/59235486/208385906-0286a075-bea5-411d-a88f-21835fea21db.png">

...but the secretStoreRef to the vault-backend seems correct. The error message in the external-secrets pod is as follows:
`{"level":"error","ts":1671439463.1461892,"logger":"controllers.ExternalSecret","msg":"could not update Secret","ExternalSecret":"team-test1/`

Which is because the new operator does not have ownership over the old secrets. Given there is already a migration tool I highly recommend we use this, as otherwise we would have to delete all the old kubernetes Secrets and recreate - but for our clients this could be extremely disruptive, as control plane rate-limiting could easily bottleneck our capacity to recreate those secrets with the new operator. With the migration tool, it's generally a lot more graceful and follows a blue-green type pattern.

j-zimnowoda commented on helmfile.d/helmfile-06.init.yaml:
```diff
@@ -33,13 +33,12 @@ releases:
     values:
       - ../values/istio-operator/istio-operator-raw.gotmpl
       - snippets/common.gotmpl
-  - name: external-secrets
```
Please add removal of this Helm release to the upgrade.yaml file

Morriz replied:
It has just been moved

j-zimnowoda replied:
no, no it was moved to different namespace, therefor it is a different helm release

Morriz replied:
Ok

Morriz replied:
But I am on the road for the next three days, so please, if you would want to add that one line to this PR I would be very grateful

j-zimnowoda replied:
sure, I will do so

j-zimnowoda commented on vendor/kes-to-eso/migrate-post.sh:
```diff
@@ -0,0 +1,11 @@
+mkdir -p eso_files
+
+# Generate ESO files and apply them
+./kes-to-eso generate -i kes_files -o eso_files -n $KES_NAMESPACE
```
It seems that the binary needs to be compiled for different architecture. I put new version

Morriz replied:
Hmmm. I downloaded amd64

j-zimnowoda replied:
new version uploaded

j-zimnowoda reviewed (changes requested):
I updated the binary and simplify the script.
However I am struggling with the following error that kube config cannot be found, while performing the post hook for upgrade involving the migration

Morriz commented:
Can you log that in their repo? Is there any output?

j-zimnowoda commented:
> Can you log that in their repo? Is there any output?

I have just relied that I pasted wrong snippet. The issue is on our side. The upgrade script cannot find Kubernetes config while performing the upgrade, so the issue is on our side.

j-zimnowoda commented:
Now I see that kestoeso requires kubeconfig
```
+ ./vendor/kes-to-eso/kestoeso apply --all-secrets --all-namespaces
time="2023-01-02T13:08:21Z" level=fatal msg="stat /home/app/.kube/config: no such file or directory"
```

j-zimnowoda commented:
Upgrade went successfully 
```
Current version of otomi: 0.17.1
Upgrade records detected for version dev
operation: vendor/kes-to-eso/migrate-post.sh && echo "Success"
Upgrade from KES to ESO
Scaling down KES
deployment.apps/external-secrets scaled
Patching secrets ownership KES to ESO
Removing KES CR
externalsecret.kubernetes-client.io "aa" deleted
Uninstalling KES
release "external-secrets" uninstalled
Removing KES CRD
customresourcedefinition.apiextensions.k8s.io "externalsecrets.kubernetes-client.io" deleted
External secrets upgrade succeeded
Success

time="2023-01-02T16:41:42Z" level=info msg="Secret team-t1/aa updated successfully"
time="2023-01-02T16:41:42Z" level=info msg="Updated 1 secrets"
```

and external secrets are synced correctly
```
# k get externalsecrets.external-secrets.io -A

NAMESPACE   NAME   STORE           REFRESH INTERVAL   STATUS         READY
team-t1     aa     vault-backend   1h                 SecretSynced   True
```
